### PR TITLE
handle the case when allocSpan fails to allocate memory

### PR DESCRIPTION
**Pull request by rafzi**
see #100

### CODE REVIEW

File: Recast/Source/RecastRasterization.cpp [hunk 1]
```diff
@@ -417,8 +449,14 @@ void rcRasterizeTriangles(rcContext* ctx, const float* verts, const unsigned cha
 		const float* v1 = &verts[(i*3+1)*3];
 		const float* v2 = &verts[(i*3+2)*3];
 		// Rasterize.
-		rasterizeTri(v0, v1, v2, areas[i], solid, solid.bmin, solid.bmax, solid.cs, ics, ich, flagMergeThr);
+		if (!rasterizeTri(v0, v1, v2, areas[i], solid, solid.bmin, solid.bmax, solid.cs, ics, ich, flagMergeThr))
+		{
+			ctx->log(RC_LOG_ERROR, "rcRasterizeTriangles: Out of memory.");
+			return false;
```
Comment (Janiels) [inline Recast/Source/RecastRasterization.cpp]: The `RC_TIMER_RASTERIZE_TRIANGLES` timer should be stopped before returning.

File: Recast/Source/RecastRasterization.cpp [hunk 2]
```diff
@@ -390,10 +416,16 @@ void rcRasterizeTriangles(rcContext* ctx, const float* verts, const int /*nv*/,
 		const float* v1 = &verts[tris[i*3+1]*3];
 		const float* v2 = &verts[tris[i*3+2]*3];
 		// Rasterize.
-		rasterizeTri(v0, v1, v2, areas[i], solid, solid.bmin, solid.bmax, solid.cs, ics, ich, flagMergeThr);
+		if (!rasterizeTri(v0, v1, v2, areas[i], solid, solid.bmin, solid.bmax, solid.cs, ics, ich, flagMergeThr))
+		{
+			ctx->log(RC_LOG_ERROR, "rcRasterizeTriangles: Out of memory.");
+			return false;
```
Comment (Janiels) [inline Recast/Source/RecastRasterization.cpp]: The `RC_TIMER_RASTERIZE_TRIANGLES` timer should be stopped before returning.

File: Recast/Source/RecastRasterization.cpp [hunk 3]
```diff
@@ -362,10 +382,16 @@ void rcRasterizeTriangles(rcContext* ctx, const float* verts, const int /*nv*/,
...
```
Comment (Janiels) [inline Recast/Source/RecastRasterization.cpp]: The `RC_TIMER_RASTERIZE_TRIANGLES` timer should be stopped before returning.

File: Recast/Source/RecastRasterization.cpp [hunk 4]
```diff
@@ -335,9 +349,15 @@ void rcRasterizeTriangle(rcContext* ctx, const float* v0, const float* v1, const
 
 	const float ics = 1.0f/solid.cs;
 	const float ich = 1.0f/solid.ch;
-	rasterizeTri(v0, v1, v2, area, solid, solid.bmin, solid.bmax, solid.cs, ics, ich, flagMergeThr);
+	if (!rasterizeTri(v0, v1, v2, area, solid, solid.bmin, solid.bmax, solid.cs, ics, ich, flagMergeThr))
+	{
+		ctx->log(RC_LOG_ERROR, "rcRasterizeTriangle: Out of memory.");
+		return false;
```
Comment (Janiels) [inline Recast/Source/RecastRasterization.cpp]: The `RC_TIMER_RASTERIZE_TRIANGLES` timer should be stopped before returning.

File: Recast/Include/Recast.h [hunk 5]
```diff
@@ -790,7 +790,7 @@ void rcClearUnwalkableTriangles(rcContext* ctx, const float walkableSlopeAngle,
 ///  @param[in]		smax			The maximum height of the span. [Limit: <= #RC_SPAN_MAX_HEIGHT] [Units: vx]
 ///  @param[in]		area			The area id of the span. [Limit: <= #RC_WALKABLE_AREA)
 ///  @param[in]		flagMergeThr	The merge theshold. [Limit: >= 0] [Units: vx]
-void rcAddSpan(rcContext* ctx, rcHeightfield& hf, const int x, const int y,
+bool rcAddSpan(rcContext* ctx, rcHeightfield& hf, const int x, const int y,
```
Comment (hymerman) [inline Recast/Include/Recast.h]: Should definitely comment the return values

**Comment by rafzi** [comment]
@hymerman @Janiels Thanks for your comments and support for recast, great to see it active again.

Are you sure about stopping the timers? I remember checking similar code and they are never stopped. That makes sense to me, because the timer values are to be disregarded anyways if the build process fails.

**Comment by Janiels** [comment]
@rafzi Looks like you are right - in other places the timers are not stopped either, so what you have now is fine!
Can you squash your commits so it's ready for the merge?

**Comment by rafzi** [comment]
@Janiels done

**Comment by Janiels** [comment]
It looks good tm e. @hymerman, can you merge if it also looks good to you?

**Comment by hymerman** [comment]
I think those other places are wrong - the timer should be stopped before returning. IIRC I noticed this problem a long time ago - what I did was create a little RAII wrapper for profile scopes so I didn't have to worry about ending them. I wasn't allowed to contribute changes back then, but I'll write the code again when I get time. For now, could you please close them @rafzi, and if you've time could you point out the places that don't do this?

**Comment by rafzi** [comment]
@hymerman I agree, I was just going for consistency.

It's not right to just sprinkle stop calls everywhere - this is C++ after all.

As the changes are unrelated I made a separate PR: #137

**Comment by hymerman** [comment]
Great stuff! I'll merge in #137 first then this one.

**Comment by hymerman** [comment]
Nice trivial merge :) If you're up to it though could you rebase instead?

**Comment by rafzi** [comment]
@hymerman Well I created a conflict for myself, but it's rebased now. :)

**Comment by grahamboree** [comment]
Looks good to me. :+1:

**Comment by hymerman** [comment]
Good stuff, thanks again @rafzi!

**Pull request merged**